cpp/src/io/orc/writer_impl.cu

```diff
@@ -1121,82 +1147,151 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
-  auto const num_skip                = is_granularity_rowgroup ? 0 : num_rowgroup_blobs;
-  gpu::orc_encode_statistics(blobs.device_ptr(),
-                             stat_merge.device_ptr(num_skip),
-                             stat_chunks.data() + num_skip,
-                             num_stat_blobs - num_skip,
-                             stream);
-  stat_merge.device_to_host(stream);
-  blobs.device_to_host(stream, true);
-
+  // we have to encode the row groups now IF they are being written out
   auto rowgroup_blobs = [&]() -> std::vector<ColStatsBlob> {
     if (not is_granularity_rowgroup) { return {}; }
+
+    hostdevice_vector<uint8_t> blobs =
+      allocate_and_encode_blobs(rowgroup_merge, rowgroup_chunks, num_rowgroup_blobs, stream);
+
     std::vector<ColStatsBlob> rowgroup_blobs(num_rowgroup_blobs);
     for (size_t i = 0; i < num_rowgroup_blobs; i++) {
-      auto const stat_begin = blobs.host_ptr(rowgroup_stat_merge[i].start_chunk);
-      auto const stat_end   = stat_begin + rowgroup_stat_merge[i].num_chunks;
+      auto const stat_begin = blobs.host_ptr(rowgroup_merge[i].start_chunk);
+      auto const stat_end   = stat_begin + rowgroup_merge[i].num_chunks;
       rowgroup_blobs[i].assign(stat_begin, stat_end);
     }
     return rowgroup_blobs;
   }();
 
+  return {std::move(rowgroup_blobs),
+          std::move(stripe_chunks),
+          std::move(stripe_merge),
+          std::move(col_stats_dtypes),
+          std::move(col_types)};
+}
+
+writer::impl::encoded_footer_statistics writer::impl::finish_statistic_blobs(
+  int num_stripes, writer::impl::persisted_statistics& per_chunk_stats)
+{
+  auto stripe_size_iter = thrust::make_transform_iterator(per_chunk_stats.stripe_stat_merge.begin(),
+                                                          [](auto const& i) { return i.size(); });
+
+  auto const num_columns = per_chunk_stats.col_types.size();
+  auto const num_stripe_blobs =
+    thrust::reduce(stripe_size_iter, stripe_size_iter + per_chunk_stats.stripe_stat_merge.size());
+  auto const num_file_blobs = num_columns;
+  auto const num_blobs      = static_cast<int>(num_stripe_blobs + num_file_blobs);
+
+  if (num_stripe_blobs == 0) { return {}; }
+
+  // merge the stripe persisted data and add file data
+  rmm::device_uvector<statistics_chunk> stat_chunks(num_blobs, stream);
+  hostdevice_vector<statistics_merge_group> stats_merge(num_blobs, stream);
+
+  // we need to merge the stat arrays from the persisted data.
+  // this needs to be done carefully because each array can contain
+  // a different number of stripes and stripes from each column must be
+  // located next to each other. We know the total number of stripes and
+  // we know the size of each array. The number of stripes per column in a chunk array can
+  // be calculated by dividing the number of chunks by the number of columns.
+  // That many chunks need to be copied at a time to the proper destination.
+  size_t offset = 0;
```

Comment:
I'm wondering if we might choose a better name for `offset`.

Reply:
Changed it to `num_entries_seen`. Do you have a better suggestion? I didn't feel strongly about this name, but couldn't find a better one.